FED-178 Add Documentation for `uiJsComponent` (#769)
Workiva/over_react: doc/wrapping_js_components.md

- Create a file for your component. The file name does not have to be unique for this process and can be consistent with existing naming patterns in your package.

### Adding the JS Variable

The first step is to create the interop getter to make the JS implementation accessible in Dart. The code written for this section is shown below.

```dart
@JS()
library some_library.path.to.component;

import 'package:js/js.dart';

@JS('MagicalJsPackage.ArbitraryComponent')
external ReactClass get _jsArbitraryComponent;
```

There are three pieces here:

1. Adding directives
1. Creating the getter
1. Attaching the annotation

#### Adding the Directives

To get started, add the important directives to the top of the file.

```dart
@JS()
library some_library.path.to.component;

import 'package:js/js.dart';
```

The `js.dart` import gives access to the `JS()` annotation. The `JS()` annotation is necessary because the file is defining interop APIs. Lastly, the `library` directive is necessary because the `JS()` annotation must be attached to a specific library.

#### Creating the Getter

Creating the getter is very simple.

```dart
external ReactClass get _jsArbitraryComponent;
```

> The code here can basically be copied and pasted from the code found in this doc, just changing the name of the getter to be something more specific than `_jsArbitraryComponent`. However, the name can really be anything. We have the norm of using `_js{ComponentName}` because:

Reviewer: #nit In the context of stylistic programming guidelines, "convention" would be more standard way to refer to this than "norm".
```suggestion
The code here can basically be copied and pasted from the code found in this doc, just changing the name of the getter to be something more specific than `_jsArbitraryComponent`. However, the name can really be anything. We have the convention of using `_js{ComponentName}` because:
```

```dart
@JS('ArbitraryComponent')
```

then Dart will expect to find the property `ArbitraryComponent` on the `window`. However, we **strongly** recommend ensuring that your bundle is namespaced and all exports are nested inside that namespace. This is important to ensure that there are not naming collisions due to two components with the same name being attached to the `Window`. Consequently, if the package is named `MagicalJsPackage`, your annotation would probably look like:

```dart
@JS('MagicalJsPackage.ArbitraryComponent')
```

> Reminder: The JavaScript bundle for the JS must be loaded into the application, [potentially similar to how React is added][react-script], or else this interop step will always fail at runtime.

To check that the name is correct, you can use the browser devtools to attempt to access the property on the `Window`. If the property is valid and generally where you expect it to be then it's likely correct.

> <img src='./images/ui_js_component/interop_ref.gif' alt='Accessing the property in the devtools' />

Reviewer: Don't forget to optimize these images and rewrite the history.

1. The prop's function signature is not the same as the version on `UbiquitousDomPropsMixin`. This typically takes the form of adding extra parameters that the standard event handlers do not have.

This causes problems in Dart because if the type of the function signature does not match what Dart expects, it will throw a runtime error. Here's what that could look like in TS:

```ts
interface ArbitraryComponentProps {
  // note that the `id` parameter isn't usually part of the `onClick` handler signature
  onClick: (e: React.MouseEvent, id: string) => void;
}
```

In Dart, to work around that, you would do something like:

```dart
@Props(keyNamespace: '')
mixin ArbitraryComponentProps on UiProps {
  @Accessor(key: 'onClick')
  void Function(SyntheticMouseEvent event, String id) onClick_ArbitraryComponent;

  @Deprecated('Use onClick_ArbitraryComponent for proper typing')
  @override
  get onClick;

  @Deprecated('Use onClick_ArbitraryComponent for proper typing')
  @override
  set onClick(value);
}
```

The steps are:

1. Add overridden getters and setters for the prop. Overriding is necessary because `UiProps` already contributes the property.

> 1. Deprecate those getters and setters. This isn't necessary to avoid runtime errors, but it's an important step because those properties should not be accessed by devs.

Reviewer: ```suggestion
> This isn't necessary to avoid runtime errors
```
I think the opposite is true. If consumers set callbacks that don't match the signature the JS code calls them with, you'll get a runtime error.

This section will give a basic example of what writing this mixin looks like and then give additional examples of more complex type conversions.

#### Props Conversion Example

Start by adding your component imports to your file, if you haven't already.

```dart
import 'package:over_react/over_react.dart';

part 'arbitrary_component.over_react.g.dart';
```

These are just the standard OverReact component directives. Altogether, the imports section of your file should look something like:

```dart
@JS()
library some_library.path.to.component;

import 'package:js/js.dart';
import 'package:over_react/over_react.dart';

part 'arbitrary_component.over_react.g.dart';
```

After that, move on to converting the TS props interface to a Dart mixin. Say you are given a props interface (or type) that looks like:

```ts
interface ArbitraryComponentProps {
  anObjProp: Record<string, bool>;
  content: string;

>   onClick: (e: React.MouseEvent, id: string) => void;

Reviewer 2: ```suggestion
  onClick: (e: React.SyntheticMouseEvent, id: string) => void;
```

Reviewer 2: Also - this "intricate" conversion seems like something that may be unique to MUI JS... Should we mention that if the JS component just uses the HTML `onClick` function signature - that the namespacing is not necessary?

Reviewer 2: Or maybe include a dom callback like `onBlur` as well - that doesn't require a custom conversion and be really explicit that the `onClick` used in the example is for demonstration purposes only - and that most JS component libs don't customize dom prop callbacks like this?

Author: In TS, is it `React.SyntheticMouseEvent`? I only saw [types like these](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5a03f363621433711005972e084f8e2111662299/types/react/v17/index.d.ts#L1156-L1160), where the name is just something like `ClipboardEvent` and it extends from `SyntheticEvent` (or `UiEvent`, which extends from `SyntheticEvent`).

I agree though that we should specific this is a rare case. I'll update that!

Reviewer 2: mmm.... you might be right that its not a synthetic event in dom callback cases in typescript land.

Author: Ohh I may be misunderstanding, but I don't think that's quite what I was saying. I meant that in TS, synthetic events don't have the word "synthetic" in them. A `React.MouseEvent` will map to a Dart `SyntheticMouseEvent`

Usually OverReact uses APIs that obfuscate the need to create a factory proxy, but in this process it is a manual step. However, all it means is that when you invoke `uiJsComponent`, you'll pass in a `ReactJsComponentFactoryProxy` that wraps a raw JS component.

### Level 1: JavaScript Land

In order to have `ReactJsComponentFactoryProxy` wrap a JS component though, you need to be able to access your JS component in Dart. You will do this by using JS interop APIs to create a Dart getter that references the JS implementation.

```dart
@JS('MagicalJsPackage.ArbitraryComponent')
external ReactClass get _jsArbitraryComponent;
```

That getter can then be used whenever it's necessary to reference the raw JS component, such as when instantiating `ReactJsComponentFactoryProxy`.

In order for that getter to be valid though, it must be pointing to tangible JavaScript somewhere. The interop APIs look for that JavaScript as a property of the `window`. Your JavaScript component can be included in several ways, but it must happen before the `UiFactory` attempts to reference the interop getter. Otherwise, a runtime error will fire.

> One way to load the JavaScript is to use a bundle like [React usually is][react-dart]. If you are wrapping a JS library without any customizations, it's likely the bundle already exists as a [UMD module](https://jameshfisher.com/2020/10/04/what-are-umd-modules/) or similar format. If you need to write your own JS or make customizations, it's necessary to create and load a custom build. If you have any questions around this, let us know in [Slack][slack]!

Reviewer: > One way to load the JavaScript is to use a bundle like [React usually is][react-dart].

I think there's a wording issue here. Maybe something like this?

> One way to load the JavaScript is to use a bundle, like how [React is loaded in Dart][react-dadt].

```dart
@Props(keyNamespace: '')
mixin ArbitraryComponentProps on UiProps {
  Map get aMapProp => unjsifyMapProp(_aMapProp$rawJs);

  set aMapProp(Map value) => _aMapProp$rawJs = jsifyMapProp(value);

  @Accessor(key: 'aMapProp')
  JsMap _aMapProp$rawJs;
}
```

> **For more information** on why getters and setters are used here, see the [Conversion through Getters and Setters section](#conversion-through-getters-and-setters).

The steps are:

1. Create a `JsMap` property that will represent the raw JS data for the component. This is necessary because the raw JS is not in a consumable form and should not be accessed directly.
1. Add an `@Accessor` annotation to the property with a key that matches the prop name. This will link the `JsMap` property to the JS prop.
1. Add `Map` getters and setters named the same as the prop.
   - The getter should use `unjsifyMapProp` to convert the JS property to a Dart `Map`
   - The setter should use `jsifyMapProp` to convert the Dart `Map` to JS.

###### Converting Refs

> If you read the "Index Signature" section of [Converting Maps](#converting-maps), this section will be very similar (just with different interop utilities). Say you have props that look like:

Reviewer: This link is broken/outdated; I think it should be:
```suggestion
If you read the "Index Signature" section of [Converting Javascript Object Types](#converting-javascript-object-types), this section will be very similar (just with different interop utilities). Say you have props that look like:
```
